atc/creds/pool.go

```diff
@@ -0,0 +1,107 @@
+package creds
+
+import (
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	"encoding/json"
+)
+
+//go:generate counterfeiter . VarSourcePool
+
+type VarSourcePool interface {
+	FindOrCreate(lager.Logger, map[string]interface{}, ManagerFactory) (Secrets, error)
+}
+
+type inPoolManager struct {
+	manager        Manager
+	secretsFactory SecretsFactory
+	lastUseTime    time.Time
+}
+
+func (m *inPoolManager) Close(logger lager.Logger) {
+	m.manager.Close(logger)
+}
+
+func (m *inPoolManager) NewSecrets() Secrets {
+	m.lastUseTime = time.Now()
+	return m.secretsFactory.NewSecrets()
+}
+
+type varSourcePool struct {
+	pool map[string]*inPoolManager
+	lock sync.Mutex
+	ttl  time.Duration
+}
+
+func (pool *varSourcePool) FindOrCreate(logger lager.Logger, config map[string]interface{}, factory ManagerFactory) (Secrets, error) {
+	b, err := json.Marshal(config)
+	if err != nil {
+		return nil, err
+	}
+
+	key := string(b)
+
+	pool.lock.Lock()
+	defer pool.lock.Unlock()
+
+	if _, ok := pool.pool[key]; !ok {
+		manager, err := factory.NewInstance(config)
+		if err != nil {
+			return nil, err
+		}
+		err = manager.Init(logger)
+		if err != nil {
+			return nil, err
+		}
+		secretsFactory, err := manager.NewSecretsFactory(logger)
+		if err != nil {
+			return nil, err
+		}
+
+		pool.pool[key] = &inPoolManager{
+			manager:        manager,
+			secretsFactory: secretsFactory,
+		}
+	} else {
+		logger.Debug("found-existing-credential-manager")
+	}
+
+	return pool.pool[key].NewSecrets(), nil
+}
+
+func (pool *varSourcePool) Collect(logger lager.Logger) error {
+	pool.lock.Lock()
+	defer pool.lock.Unlock()
+
+	logger.Debug("before collect", lager.Data{"pool-size": len(pool.pool)})
```

Comment:
this should be done using a session like so:

```go
logger = logger.Session("collect", lager.Data{"pool-size": len(pool.pool)})

logger.Debug("start")
defer logger.Debug("done")
```

also, log messages should be hyphen-delimited, though in the above example that becomes a moot point. it's kind of weird but it's the [`lager`](https://github.com/cloudfoundry/lager) way.

Reply:
@vito We cannot put `len(pool.pool)` to session, because pool size might change after collecting. The reason why I add pool-size in logs is to track pool size before and after collect.

Also, when `Collect` is called, the runner has started a session and only passes session logger to `Collect`, so I don't think we should start a session again in `Collect`.

I have changed log actions to hyphen-delimited.